clippy_lints/src/methods/unnecessary_map_or.rs

```diff
@@ -0,0 +1,119 @@
+use clippy_config::msrvs::{self, Msrv};
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::eager_or_lazy::switch_to_eager_eval;
+use clippy_utils::source::{snippet, snippet_opt};
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::is_local_used;
+use clippy_utils::{get_parent_expr, path_to_local_id};
+use rustc_ast::util::parser::AssocOp;
+use rustc_ast::LitKind::Bool;
+use rustc_errors::Applicability;
+use rustc_hir::{BinOpKind, Expr, ExprKind, PatKind};
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::UNNECESSARY_MAP_OR;
+
+pub(super) enum Variant {
+    Ok,
+    Some,
+}
+impl Variant {
+    pub fn variant_name(&self) -> &'static str {
+        match self {
+            Variant::Ok => "Ok",
+            Variant::Some => "Some",
+        }
+    }
+
+    pub fn method_name(&self) -> &'static str {
+        match self {
+            Variant::Ok => "is_ok_and",
+            Variant::Some => "is_some_and",
+        }
+    }
+}
+
+// Only checking map_or for now
+pub(super) fn check(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    recv: &Expr<'_>,
+    def: &Expr<'_>,
+    map: &Expr<'_>,
+    msrv: &Msrv,
+) {
+    if let ExprKind::Lit(def_kind) = def.kind
+        && let typeck_results = cx.typeck_results()
+        && let recv_ty = typeck_results.expr_ty(recv)
+        && (is_type_diagnostic_item(cx, recv_ty, sym::Option) || is_type_diagnostic_item(cx, recv_ty, sym::Result))
+        && let Bool(def_bool) = def_kind.node
+    {
+        let variant = if is_type_diagnostic_item(cx, recv_ty, sym::Option) {
+            Variant::Some
+        } else {
+            Variant::Ok
+        };
+
+        let (sugg, method) = if let ExprKind::Closure(map_closure) = map.kind
+        && let closure_body = cx.tcx.hir().body(map_closure.body)
+        && let closure_body_value = closure_body.value.peel_blocks()
+        && let ExprKind::Binary(op, l, r) = closure_body_value.kind
+        && let Some(param) = closure_body.params.first()
+        && let PatKind::Binding(_, hir_id, _, _) = param.pat.kind
+        // checking that map_or is one of the following:
+        // .map_or(false, |x| x == y)
+        // .map_or(false, |x| y == x) - swapped comparison
+        // .map_or(true, |x| x != y)
+        // .map_or(true, |x| y != x) - swapped comparison
+        && ((BinOpKind::Eq == op.node && !def_bool) || (BinOpKind::Ne == op.node && def_bool))
+        && let non_binding_location = if path_to_local_id(l, hir_id) { r } else { l }
+        && switch_to_eager_eval(cx, non_binding_location)
+        // xor, because if its both then thats a strange edge case and
+        // we can just ignore it, since by default clippy will error on this
+        && (path_to_local_id(l, hir_id) ^ path_to_local_id(r, hir_id))
+        && !is_local_used(cx, non_binding_location, hir_id)
+        && typeck_results.expr_ty(l) == typeck_results.expr_ty(r)
```

Comment:
Small formatting change, surprised rustfmt didn't catch this
```suggestion
            && let closure_body = cx.tcx.hir().body(map_closure.body)
            && let closure_body_value = closure_body.value.peel_blocks()
            && let ExprKind::Binary(op, l, r) = closure_body_value.kind
            && let Some(param) = closure_body.params.first()
            && let PatKind::Binding(_, hir_id, _, _) = param.pat.kind
            // checking that map_or is one of the following:
            // .map_or(false, |x| x == y)
            // .map_or(false, |x| y == x) - swapped comparison
            // .map_or(true, |x| x != y)
            // .map_or(true, |x| y != x) - swapped comparison
            && ((BinOpKind::Eq == op.node && !def_bool) || (BinOpKind::Ne == op.node && def_bool))
            && let non_binding_location = if path_to_local_id(l, hir_id) { r } else { l }
            && switch_to_eager_eval(cx, non_binding_location)
            // xor, because if its both then thats a strange edge case and
            // we can just ignore it, since by default clippy will error on this
            && (path_to_local_id(l, hir_id) ^ path_to_local_id(r, hir_id))
            && !is_local_used(cx, non_binding_location, hir_id)
            && typeck_results.expr_ty(l) == typeck_results.expr_ty(r)
```